feat: integrate shepherd output for contile

tiftran opened the pull request:
main things that in the changes to accommodate the shepherd output:

click_hosts, impression_hosts have been removed from advertiser specific level
DEFAULTS have been moved to their own settings config env
at the advertiser specific level is where the country info lies now

Closes [CONSVC-1807](https://mozilla-hub.atlassian.net/jira/software/c/projects/CONSVC/issues/CONSVC-1807)

tiftran commented on src/adm/settings.rs:
```diff
@@ -408,45 +514,73 @@ impl From<&mut Settings> for HandlerResult<AdmFilter> {
+        let excluded_countries_200 = settings.excluded_countries_200;
+
+        let mut settings_str = "".to_string();
+        if Path::new(&settings.adm_settings).exists() {
+            if let Ok(mut f) = File::open(&settings.adm_settings) {
+                settings_str = f
+                    .read_to_string(&mut settings.adm_settings)
+                    .map_err(|e| {
+                        HandlerError::internal(&format!(
+                            "Could not read {}: {:?}",
+                            settings.adm_settings, e
+                        ))
+                    })?
+                    .to_string();
```
this feels jank... im not sure how to not make it feel less jank

jrconlin commented on src/adm/settings.rs:
```diff
@@ -112,41 +126,20 @@ pub struct PathFilter {
-    pub(crate) include_regions: Vec<String>,
-    pub(crate) ignore_advertisers: Option<Vec<String>>,
-    pub(crate) ignore_dmas: Option<Vec<u8>>,
+    // TODO: handle country and path parsing.
```
Can probably remove this TODO

jrconlin commented on src/adm/settings.rs:
```diff
@@ -244,62 +265,171 @@ impl Serialize for AdmFilterSettings {
+                ConfigError::Message(format!("Invalid advertiser info {}", &advertiser))
+            })? {
+                // Delete allows us to delete this advertiser. It may appear at the same
+                // level as the country listsings.
```
s/listsings/listings/

jrconlin commented on src/adm/settings.rs:
```diff
@@ -244,62 +265,171 @@ impl Serialize for AdmFilterSettings {
+                }
+                // if it's not a `delete`, than it's part of the
+                // list of per country advertiser filters.
+                //*
```
Let's remove the comment block bits here. We use this code.

jrconlin commented on src/adm/settings.rs:
```diff
@@ -319,11 +449,10 @@ impl AdmFilterSettings {
             .map_err(|e| ConfigError::Message(format!("Could not download settings: {:?}", e)))?;
-        let mut reply =
-            AdmFilterSettings::try_from(String::from_utf8(contents).map_err(|e| {
+        let reply =
```
kinda wondering if we could collapse this into the `OK()`

e.g.
``` 
        Ok(AdmFilter::advertisers_from_string(&String::from_utf8(contents).map_err(|e| {
                ConfigError::Message(format!("Could not read ADM Settings: {:?}", e))
            })?)?)
```

jrconlin commented on src/adm/settings.rs:
```diff
@@ -408,45 +521,68 @@ impl From<&mut Settings> for HandlerResult<AdmFilter> {
             .to_lowercase();
-        let mut all_include_regions = HashSet::new();
+        // No longer needed since each advertiser has it's own set of countries.
+        // let mut all_include_regions = HashSet::new();
```
Since it's no longer needed, let's drop this.

jrconlin commented on src/adm/settings.rs:
```diff
@@ -455,10 +591,10 @@ impl From<&mut Settings> for HandlerResult<AdmFilter> {
 
     use super::*;
-    use crate::web::test::adm_settings;
+    //use crate::web::test::adm_settings;
```
And remove the commented out `// use` statements.

jrconlin commented on src/web/test.rs:
```diff
@@ -108,11 +120,12 @@ impl MockAdm {
             .into_owned()
-            .collect()
+            .collect();
+        bits
```
since we're no longer debugging this, let's just return 
```
Url::parse(&format!("{}?{}", self.endpoint_url, query_string))
            .expect("Couldn't parse request_rx result")
            .query_pairs()
            .into_owned()
            .collect()
```

jrconlin commented on test-engineering/contract-tests/volumes/client/scenarios.yml:
```diff
@@ -231,297 +231,297 @@ scenarios:
-                      value: 'json'
-                    - name: results
-                      value: '5'
+#  - name: legacy_image_filter
```
We should drop this or at least add a comment block indicating that we no longer need to do image filter tests, since they're now part of the default configuration. 

Make sure to also flag context-services-test-eng as a reviewer since we're modifying the tests.

tiftran replied:
eek, lemme go through those tests and uncomment accordinly first, I initially commented out to help me test locally.

jrconlin commented on test-engineering/contract-tests/README.md:
```diff
@@ -17,9 +17,7 @@ the individual components of the suite.
-docker compose \ 
--f test-engineering\contract-tests\docker-compose.yml \ 
-up --abort-on-container-exit --build
+u
```
Guessing this is a typo? Probably restore what was there.

tiftran replied:
eeek oops

jrconlin reviewed (changes requested):
A few minor fix ups, but looks good. Thanks!

jrconlin commented on src/adm/filter.rs:
```diff
@@ -45,22 +45,22 @@ lazy_static! {
 /// In addition there is a special `DEFAULT` value which is a filter
```
Let's delete references to the `DEFAULT` since it's no longer present.

jrconlin reviewed (changes requested):
Looks pretty good. A few minor fixups to remove some older comments and collapse some code.

Trinaa reviewed (changes requested):
I believe that there are contract tests being excluded, for instance `success_tiles_cache_expired`, that are intended to test functionality unrelated to this change. Perhaps we can discuss and adjust the scenarios accordingly?

jrconlin commented:
@Trinaa: We commented out those tests initially because we knew there would be some work getting them back online. We changed around a number of the configs to fit the way that Contile actually works, rather than the wild guesses we made initially. (e.g. the Click, Image, and Impression URLs don't really change, since we own those, so they were moved out of ADM_Settings and into ADM_Defaults. We just forgot to turn those tests back on. (oops.)

I've got most of the tests working now, but there are a couple that i'm hitting that I'm not sure how to fix. One of them is the one that's currently failing. I think it's mostly a matter of trying to mesh up the inputs and responses. If you have any suggestions, I'm all ears.

Trinaa commented on test-engineering/contract-tests/docker-compose.yml:
```diff
@@ -30,9 +30,23 @@ services:
       CONTILE_ADM_SUB1: sub1_test
       CONTILE_ADM_PARTNER_ID: partner_id_test
       CONTILE_ADM_HAS_LEGACY_IMAGE: '["Example ORG","Example COM"]'
+      CONTILE_ADM_DEFAULTS: '{"image_hosts":["example.com","example.org"],"impression_hosts":["example.com","example.org"],"click_hosts":["example.com","example.org"]}'
```
This is a duplicate environment variable.

Trinaa commented on test-engineering/contract-tests/docker-compose.yml:
```diff
@@ -30,6 +30,24 @@ services:
       CONTILE_ADM_SETTINGS: /tmp/contile/adm_settings.json
+      CONTILE_ADM_DEFAULTS: '{
+        "advertiser_urls": [],
+        "click_hosts": [
+            "example.com",
+            "example.org",
+            "dunbroch.co.uk"
+        ],
+        "impression_hosts": [
+            "example.com",
+            "example.org",
+            "dunbroch.co.uk"
+        ],
```
**nit:** indentation

```suggestion
        "click_hosts": [
          "example.com",
          "example.org",
          "dunbroch.co.uk"
        ],
        "impression_hosts": [
          "example.com",
          "example.org",
          "dunbroch.co.uk"
        ],
```

Trinaa commented on test-engineering/contract-tests/volumes/contile/adm_settings.json:
```diff
@@ -1,104 +1,186 @@
-        ]
-    }
+  "Example COM": {
+    "US": [
```
**suggestion:** Would it make sense to order regions alphabetically?

Trinaa reviewed (commented):
@tiftran @jrconlin 
What do you think about adding a contract test for the new `delete` advertiser key field (could be a subsequent task depending on feasibility and effort)?

tiftran commented:
@Trinaa @jrconlin I think I want to opt to add the contract test in another PR considering this one is so big already

Trinaa reviewed (approved):
@tiftran

Contract test changes LGTM 🚀 . Let me know if/how I can support the contract test addition in subsequent PR.

pjenvey commented on src/adm/tiles.rs:
```diff
@@ -248,17 +249,22 @@ pub async fn get_tiles(
-            .take(settings.adm_max_tiles as usize)
-            .collect()
-    };
+    let mut filtered: Vec<Tile> = Vec::new();
```
Is this change necessary? Not that it's terribly important with `adm_max_tiles` being small but previously it only acquired the read lock once.

jrconlin replied:
IIRC, the problem is that `filter_and_process()` now returns a `Result<Option<...>>`. I can't just call `?` on it otherwise I'll lose the error. That said, I can break `filter` out of the loop fairly easily.

pjenvey replied:
My fault, I missed this in the diff. You could collect into `Vec<Result<..>>` but that doesn't work well with the `take`ing, so this change makes sense.

pjenvey commented on test-engineering/contract-tests/volumes/contile/adm_settings.json:
```diff
@@ -1,104 +1,186 @@
-            "CN"
-        ]
-    }
+  "Example COM": {
```
Sorry I should have suggested this much earlier, but how about adding another level of nesting to these json settings? It would make adding extra buckets of data later easier if needed.

```json
{
    "adm_advertisers": {
        "Example COM": {
             ...
        }
    },
    "last_modified": "2022-10-17 21:50:00 UTC"
}
```

E.g. a timestamp from shepherd (granted we could also use GCS's file timestamps for that)

jrconlin replied:
Let's make that a future feature, since we'll have to modify Shepherd output as well.

jrconlin commented:
Having talked with @pjenvey I think there are a few changes we should definitely consider before this gets landed. Sadly, these changes will also require changes to the Shepherd UI.

1) we should move the advertisers into a sub block (see [this comment](https://github.com/mozilla-services/contile/pull/446#pullrequestreview-1144845467)] 
2) Let's drop the `delete` flag. If an advertiser is no longer to be recognized, it should just not be part of the data. This will also mean we can use proper `serde` to read data in, so we can drop a lot of [this code](https://github.com/mozilla-services/contile/pull/446/files#diff-34387244eabf61c959f3a7e535ca04a551c779b9ee645d2e93303c0020f5a881)

If you run out of rude things to say about me, feel free to ask. I'll provide an endless supply of material you can use to insult me with.

jrconlin reviewed (changes requested):
See [this comment](https://github.com/mozilla-services/contile/pull/446#issuecomment-1282845890)

tiftran commented:
@jrconlin @pjenvey just wanted to get a check on removing delete and adding the sub level before moving onto adding back the include_regions bit?

jrconlin commented:
Yes, we want to remove the "delete" option (which should also GREATLY simplify things, since we should be able to remove [the custom parser code](https://github.com/mozilla-services/contile/pull/446/files#diff-34387244eabf61c959f3a7e535ca04a551c779b9ee645d2e93303c0020f5a881R291) and just use simple `serde` rules. We also want to put the `adm_filters` under their own section. I don't think we really care what they should be called, but basically the Company Name and filter info should not be at the top level.

tiftran commented:
@jrconlin @pjenvey so i _think_ i added the sub level, removed the manually parsing bits. Was hoping if you could look over my new commits, before i start adding back the include regions stuff? 🥺

jrconlin commented on src/adm/filter.rs:
```diff
@@ -380,14 +380,10 @@ impl AdmFilter {
-            .adm_advertisers
-            .get(&tile.name.to_lowercase())
-        {
+        match self.advertiser_filters.adm_advertisers.get(&tile.name) {
```
This does mean that "Amazon" and "amazon" are different companies. Do we want that?

jrconlin replied:
(I think we want to keep the `name.to_lowercase()` because we're normalizing the names when we add them to `adm_advertisers`.

tiftran replied:
huh i thought i added that back... will re add,

jrconlin replied:
Are we not indexing by lowercase? If so, we probably want to remove the lowercase we do in `Deserialize<'de> for AdmAdvertiserSettings`. 

Probably should add a unit test somewhere to ensure that we don't trip on this in the future.

tiftran replied:
<img width="1119" alt="Screen Shot 2022-10-26 at 11 49 04 AM" src="https://user-images.githubusercontent.com/25379936/198111092-35f4292e-f9d1-4f19-a084-51cc3381182f.png">
this is the current diff i see?

jrconlin commented on src/adm/settings.rs:
````diff
@@ -230,76 +237,131 @@ impl AdmPse {
+    /// Parse a JSON string containing the ADM settings. These will be generated by shepherd and
+    /// would have a format similar to the following:
+    /// ```json
+    /// {
````
Let's update this to reflect the new format.

jrconlin commented on src/adm/settings.rs:
````diff
@@ -333,70 +395,54 @@ impl AdmFilterSettings {
 /// This uses a JSON construct of settings, e.g.
 /// ```javascript
 /// /* for the Example Co advertiser... */
 /// {"Example": {
````
Again, let's make this comment match what we're doing, so add in the parent `adm_advertisers` key.

jrconlin commented on src/adm/settings.rs:
```diff
@@ -477,46 +553,90 @@ mod tests {
-                    .collect()
-        );
+    pub fn test_valid_path_filters() {
+        let adm_settings = r#"{"adm_advertisers":{"test-adv": {
```
probably want to pretty this so it's more readable.

jrconlin commented on adm_settings_test.json:
```diff
@@ -1,23 +1,5 @@
-},
-"Acme": {
-    "advertiser_urls": [
+{"Acme": {
```
Shouldn't this include the `adm_advertisers` parent?

jrconlin commented:
FWIW, I've had stuff like that get dropped because of a merge. I chalk it
up as "git being helpful".  Might be what happened here.

On Tue, Oct 25, 2022 at 2:57 PM Tif Tran ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/adm/filter.rs
> <https://github.com/mozilla-services/contile/pull/446#discussion_r1004998572>
> :
>
> > @@ -380,14 +380,10 @@ impl AdmFilter {
>      ) -> HandlerResult<Option<Tile>> {
>          // Use strict matching for now, eventually, we may want to use backwards expanding domain
>          // searches, (.e.g "xyz.example.com" would match "example.com")
> -        match self
> -            .advertiser_filters
> -            .adm_advertisers
> -            .get(&tile.name.to_lowercase())
> -        {
> +        match self.advertiser_filters.adm_advertisers.get(&tile.name) {
>
> huh i thought i added that back... will re add,
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mozilla-services/contile/pull/446#discussion_r1004998572>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAIXK32VCN3JU3R7Z2P2E3WFBJS3ANCNFSM6AAAAAAQDMU7V4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

tiftran commented on src/adm/filter.rs:
```diff
@@ -181,28 +187,25 @@ impl AdmFilter {
+                            self.source, e
+                        ))
+                    })?;
+            self.advertiser_filters.adm_advertisers.clear();
+            for (adv, setting) in advertiser_filters.adm_advertisers {
+                for country in setting.keys() {
+                    if !self.defaults.include_regions.contains(country) {
+                        self.defaults.include_regions.push(country.clone());
```
wasn't sure if this was the way to go about adding include_regions

pjenvey replied:
I don't think `include_regions` belongs in `AdmDefaults`, it's a bit confusing there since it should only be generated by contile. Despite the previous `DEFAULT` key supporting an `include_regions` value (which was a bit different) we didn't really need it and especially don't with shepherd's UI and the new per region advertiser settings, so it was good that you originally got rid of it.

Let's keep it as `AdmFilter::all_include_regions` (`all_`'s helpful to distinguish it from the `include_regions` settings) and as a `HashSet`. I think you can pretty much revert to most of the previous code, and this `update()` could easily `.clear()` and rebuild it.

cc @jrconlin we can probably kill `NoTilesForCountry` as well?

jrconlin replied:
I think we need to keep `NoTilesForCountry`. We have logic still in `filter_and_process` to optionally return a 200 for tiles that are in excluded countries (since those default to return 204), and `NoTilesForCountry` allows us to return that in `error.rs`.

pjenvey replied:
I see what you're getting at: The `settings.excluded_countries_200` flag works around [what's really a bug in the client](https://github.com/mozilla-services/contile/issues/284) where it treats our 204 EMPTY response as a "noop, reverting back to the existing tile set it has", so this setting makes us return a 200 OK instead, forcing the client to use an empty tile set instead.

You're right in that we've only applied this setting to that one code path and we're still rendering 204s in other cases, with `NoTilesForCountry` covering one of them.

However there is a bug with it, it basically "short circuits" the filtering process, I've made a test on another branch which it causes to fail: https://github.com/mozilla-services/contile/compare/shepherd-output...shepherd-output-filter-bug (we're returning empty tiles there instead of 1)

I think it makes sense to consider the other 204s we're still rendering as a bug (separate from this PR) but I think the right fix is passing this setting along to all the places rendering 204s where they can take it into account.

pjenvey commented on src/adm/filter.rs:
```diff
@@ -234,26 +236,23 @@ impl AdmFilter {
-            if let Some(ref paths) = filter.paths {
+        for filter in filters {
+            if host == filter.host {
+                let paths = &filter
+                    .paths
+                    .clone()
+                    .unwrap_or_else(|| [PathFilter::default()].to_vec());
```
Nit: you could avoid this clone here by keeping a static of the default:

```suggestion
                    .as_ref()
                    .unwrap_or_else(|| &DEFAULT_PATH_FILTER);
```

with this in the `lazy_static` block:
```rust
static ref DEFAULT_PATH_FILTER: Vec<PathFilter> = vec![PathFilter::default()];
```

pjenvey replied:
Nit: you could avoid this clone here by keeping a static of the default:

```suggestion
                let paths = filter
                    .paths
                    .as_ref()
                    .unwrap_or_else(|| &DEFAULT_PATH_FILTER);
```

with this in the `lazy_static` block:

```rust
static ref DEFAULT_PATH_FILTER: Vec<PathFilter> = vec![PathFilter::default()];
```

pjenvey reviewed (approved):
I'd prefer passing the `excluded_countries_200` setting to the render methods instead of storing this in the cache but no big deal, can always change it later.

Thanks @tiftran & @jrconlin!

pjenvey commented on src/adm/filter.rs:
```diff
@@ -381,34 +380,24 @@ impl AdmFilter {
         tags: &mut Tags,
         metrics: &Metrics,
-    ) -> Option<Tile> {
+    ) -> HandlerResult<Option<Tile>> {
```
This might not need to return a Result any longer (I'm guessing it was added for NoTilesForCountry)?
